### DISCUSSION TITLE
Does "freeze" have tree shaking?

### DISCUSSION THREAD

**Discussion by AuroraTea** [micropython/micropython#14322] [General]
If not now, will it be considered for support in the future?

**Comment by andrewleech** [comment]
What do you mean by "tree shaking"?

Ok, Google suggests terminology from JavaScript world for removing unreachable / dead code. 

I believe some of this is already done by the mpy compiler when it's guaranteed unreachable, eg if there's a 
```
X = const(False)
If X:
    #big code
```
I'm pretty sure that if branch is pruned.

I'd say this is only done for really certain stuff like const, anything else that's runtime changeable wouldn't be removed.

With libraries etc I don't think functions or anything are removed because the user might want to run them from repl / exec etc.

**Reply by AuroraTea** [reply]
Sorry, I didn't realize this term is specific to JavaScript. 😁

> the user might want to run them from repl / exec etc.

Indeed this kind of situation can occur. I hadn't considered it before.

The situation I am currently facing is that there is a lot of unnecessary content in the third-party library being referenced. I feel that this part accounts for a considerable portion of the final output. Manual deletion seems to be both troublesome and unsuitable for the freeze. I wonder if there is any other way.

**Reply by andrewleech** [reply]
Oh it might not be JavaScript specific, I'd just never heard of it before. 

Yeah it's a good point about third party libraries, they certainly often contain more than you need. 

For many larger packages in micropython-lib like aioble there's separate sub-packages that can be installed to limit the installation to certain features, but that's based on the code being split into separate python files to then only install some of the files. 

It would be neat to be able to provide a filter in the manifest freeze lines to strip out certain functions / variables / modules etc. 
I'm really not sure how much work it would be to implement this though...

**Reply by Josverl** [reply]
Iirc tree shaking is mostly used by JS bundles to avoid needing to download too large chunks of the planet in order to run 'hello website'. The size optimisation can be done in micropython by freezing, which avoids wasting RAM.

I don't think is possible to selectively load a module without a significant deviation from CPython, essentially breaking the concept of a module.

I agree that refactoring the problematic module in smaller independent submodules, either directly or through lazy loading, is the likely the better approach.

To understand better, what module is causing you issues?
